From 1638fe3c90934714d2e7a90ed5ecc21baeff2c27 Mon Sep 17 00:00:00 2001
From: Dhanavandhana Kannan <quic_dhanavan1@quicinc.com>
Date: Mon, 13 Jan 2025 20:29:27 +0530
Subject: [PATCH] UPSTREAM:CP:hostapd: Set WDS flag for all link STA

Currently when the primary TQM link is removed via ML
reconfig enabled with primary UMAC migration in
WDS connection, STA gets disconnected because WDS flag
has been set only for link STA which is connected with
primary link. When the primary TQM link is removed, it
results in the removal of whole VLAN interface.

Fix is to set WDS flag for all link STA to ensure that
when any of the link is removed, it should not result in
bringing down the whole interface.

Patch-dependency: r03-029-hostapd-Enable-WDS-AP-MLO-Support.patch
Patch-work: None

Signed-off-by: Dhanavandhana Kannan <quic_dhanavan1@quicinc.com>
---
 src/ap/ieee802_11.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 5222cd3..90eb39c 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7046,6 +7046,22 @@ static void hostapd_ml_handle_assoc_cb(struct hostapd_data *hapd,
 #endif /* CONFIG_IEEE80211BE */
 }
 
+static void set_wds_sta_flag(struct hostapd_data *hapd, struct
+			     sta_info *sta)
+{
+	struct sta_info *psta;
+	struct hostapd_data *phapd;
+
+	if (ap_sta_is_mld(hapd, sta)) {
+		for_each_mld_link(phapd, hapd) {
+			if (phapd == hapd)
+				continue;
+			psta = ap_get_sta(phapd, sta->addr);
+			if (psta)
+				psta->flags |= WLAN_STA_WDS;
+		}
+	}
+}
 
 static void handle_assoc_cb(struct hostapd_data *hapd,
 			    const struct ieee80211_mgmt *mgmt,
@@ -7158,6 +7174,7 @@ static void handle_assoc_cb(struct hostapd_data *hapd,
 			   MAC2STR(sta->addr));
 		sta->pending_wds_enable = 0;
 		sta->flags |= WLAN_STA_WDS;
+		set_wds_sta_flag(hapd, sta);
 		if (hapd->conf->mld_ap && hostapd_get_wds_mld_sta_uid(hapd, sta) < 0) {
 			wpa_printf(MSG_DEBUG, "No room for uid"
 				   "to enable 4-address WDS mode for STA "
@@ -7546,6 +7563,7 @@ void ieee802_11_rx_from_unknown(struct hostapd_data *hapd, const u8 *src,
 					"mode for STA with id %u flags 0x%x", aid, sta->flags);
 
 			sta->flags |= WLAN_STA_WDS;
+			set_wds_sta_flag(hapd, sta);
 			ret = hostapd_set_wds_sta(hapd, ifname_wds,
 						  sta->addr, aid, 1);
 			if (!ret)
-- 
2.34.1

